diff --git a/source/hostap-2.10/src/drivers/nl80211_copy.h b/source/hostap-2.10/src/drivers/nl80211_copy.h
index 4693565..7abdfc6 100644
--- a/source/hostap-2.10/src/drivers/nl80211_copy.h
+++ b/source/hostap-2.10/src/drivers/nl80211_copy.h
@@ -1481,6 +1481,7 @@ enum nl80211_commands {
 
 	NL80211_CMD_ASSOC_COMEBACK,
 
+    NL80211_CMD_SEND_USR_PACKET = 246,
 	/* add new commands above here */
 
 	/* used to define NL80211_CMD_MAX below */
