##########################################
Date: Apr 22, 2022 10:47 PM
From: 53841f78c274c28c1d57bfb2159ff7e0f61ac0ba
Subject: Supported/basic rates per VAP configuration for CAC 
Source: Comcast
License: BSD 
Upstream-Status: Pending
Signed-off-by: Tamilarasi Ganesan <tamilarasi_ganesan@comcast.com>
##########################################
diff --git a/source/hostap-2.9/src/ap/hostapd.h b/source/hostap-2.9/src/ap/hostapd.h
index e1af0bc..e1d0eb2 100755
--- a/source/hostap-2.9/src/ap/hostapd.h
+++ b/source/hostap-2.9/src/ap/hostapd.h
@@ -488,7 +488,8 @@ struct hostapd_iface {
 	 * current_mode->channels */
 	int num_rates;
 	struct hostapd_rate_data *current_rates;
-	int *basic_rates;
+	struct hostapd_rate_data *current_cac_rates;
+    int *basic_rates;
 	int freq;
 
 	u16 hw_flags;
diff --git a/source/hostap-2.9/src/ap/ieee802_11.c b/source/hostap-2.9/src/ap/ieee802_11.c
index d775382..0407878 100755
--- a/source/hostap-2.9/src/ap/ieee802_11.c
+++ b/source/hostap-2.9/src/ap/ieee802_11.c
@@ -89,7 +89,7 @@ u8 * hostapd_eid_supp_rates(struct hostapd_data *hapd, u8 *eid)
 	u8 *pos = eid;
 	int i, num, count;
 
-	if (hapd->iface->current_rates == NULL)
+	if ((hapd->iface->current_rates == NULL) && (hapd->iface->current_cac_rates == NULL))
 		return eid;
 
 	*pos++ = WLAN_EID_SUPP_RATES;
@@ -105,16 +105,27 @@ u8 * hostapd_eid_supp_rates(struct hostapd_data *hapd, u8 *eid)
 	}
 
 	*pos++ = num;
-	for (i = 0, count = 0; i < hapd->iface->num_rates && count < num;
-	     i++) {
-		count++;
-		*pos = hapd->iface->current_rates[i].rate / 5;
-		if (hapd->iface->current_rates[i].flags & HOSTAPD_RATE_BASIC)
-			*pos |= 0x80;
-		pos++;
-	}
+        if(hapd->iface->current_cac_rates) {
+                for (i = 0, count = 0; i < hapd->iface->num_rates && count < num;
+                                i++) {
+                        count++;
+                        *pos = hapd->iface->current_cac_rates[i].rate / 5;
+                        if (hapd->iface->current_cac_rates[i].flags & HOSTAPD_RATE_BASIC)
+                                *pos |= 0x80;
+                        pos++;
+                }
+        } else {
+                for (i = 0, count = 0; i < hapd->iface->num_rates && count < num;
+                                i++) {
+                        count++;
+                        *pos = hapd->iface->current_rates[i].rate / 5;
+                        if (hapd->iface->current_rates[i].flags & HOSTAPD_RATE_BASIC)
+                                *pos |= 0x80;
+                        pos++;
+                }
+        }
 
-	if (hapd->iconf->ieee80211n && hapd->iconf->require_ht && count < 8) {
+        if (hapd->iconf->ieee80211n && hapd->iconf->require_ht && count < 8) {
 		count++;
 		*pos++ = 0x80 | BSS_MEMBERSHIP_SELECTOR_HT_PHY;
 	}
