index 7893da4..865704d 100644
--- git.orig/source/broadband/webpa_parameter.c
+++ git/source/broadband/webpa_parameter.c
@@ -83,6 +83,11 @@ void getValues(const char *paramName[], const unsigned int paramCount, int index
         {
             break;
         }
+        if ((paramCount == 1) && (parameterName[(strlen(parameterName)-1)] == '.'))
+        {
+            count = 1;
+            isLargeWildCard = 0;
+        }
         WalPrint("parameterName: %s count: %d\n",parameterName,count);
         for(i = 0; i < count; i++)
         {
@@ -96,7 +101,7 @@ void getValues(const char *paramName[], const unsigned int paramCount, int index

     if(error != 1)
     {
-        isLargeWildCard = 0;
+        //isLargeWildCard = 0;
         WalPrint("compCount : %d paramCount: %d\n",compCount,paramCount);
         if(compCount > paramCount)
         {