##########################################
Date: Apr 16, 2024 1:00 PM
From: 
Subject: Add RIFS mode parameter
Source: Comcast
License: BSD
Upstream-Status: Pending
Signed-off-by: Bogdan Bogush <bogdan.bogush@comcast.com>
##########################################
diff --git a/source/hostap-2.11/src/ap/ap_config.h b/source/hostap-2.11/src/ap/ap_config.h
index de02dda..f1f2319 100644
--- a/source/hostap-2.11/src/ap/ap_config.h
+++ b/source/hostap-2.11/src/ap/ap_config.h
@@ -1105,6 +1105,7 @@ struct hostapd_config {
 	u8 vht_oper_centr_freq_seg0_idx;
 	u8 vht_oper_centr_freq_seg1_idx;
 	u8 ht40_plus_minus_allowed;
+        u8 ht_rifs;
 
 	/* Use driver-generated interface addresses when adding multiple BSSs */
 	u8 use_driver_iface_addr;
diff --git a/source/hostap-2.11/src/ap/ieee802_11_ht.c b/source/hostap-2.11/src/ap/ieee802_11_ht.c
index f90f125..b93f449 100644
--- a/source/hostap-2.11/src/ap/ieee802_11_ht.c
+++ b/source/hostap-2.11/src/ap/ieee802_11_ht.c
@@ -96,6 +96,10 @@ u8 * hostapd_eid_ht_operation(struct hostapd_data *hapd, u8 *eid)
 
 	oper->primary_chan = hapd->iconf->channel;
 	oper->operation_mode = host_to_le16(hapd->iface->ht_op_mode);
+
+	if (hapd->iconf->ht_rifs == 1)
+		oper->ht_param |= HT_INFO_HT_PARAM_RIFS_MODE;
+
 	if (hapd->iconf->secondary_channel == 1)
 		oper->ht_param |= HT_INFO_HT_PARAM_SECONDARY_CHNL_ABOVE |
 			HT_INFO_HT_PARAM_STA_CHNL_WIDTH;
