diff --git a/source/hostap-2.11/src/ap/ieee802_11.c b/source/hostap-2.11/src/ap/ieee802_11.c
index 336fec1..d620a2c 100644
--- a/source/hostap-2.11/src/ap/ieee802_11.c
+++ b/source/hostap-2.11/src/ap/ieee802_11.c
@@ -5045,7 +5045,12 @@ u16 send_assoc_resp(struct hostapd_data *hapd, struct sta_info *sta,
 #endif /* CONFIG_IEEE80211AX */
 
 	p = hostapd_eid_ext_capab(hapd, p, false);
-	p = hostapd_eid_bss_max_idle_period(hapd, p, sta->max_idle_period);
+	if (sta) {
+		p = hostapd_eid_bss_max_idle_period(hapd, p, sta->max_idle_period);
+	}else {
+		p = hostapd_eid_bss_max_idle_period(hapd, p, 0);
+		wpa_printf(MSG_DEBUG, "sta is NULL, set sta->max_idle_period to 0 ");
+	}
 	if (sta && sta->qos_map_enabled)
 		p = hostapd_eid_qos_map_set(hapd, p);
 
